### ISSUE TITLE
Implement Gradient Noise Scale monitoring

### ISSUE THREAD

**Issue by ibeltagy** [bigscience-workshop/Megatron-DeepSpeed#33]
Follow appendix A.1 https://arxiv.org/pdf/1812.06162.pdf to implement monitoring of gradient noise scale and add it to the tensorboard log.

**Comment by lintangsutawika** [comment]
I can take a look. I've already read the appendix.

**Comment by lintangsutawika** [comment]
I think I have found where each the parameters of from each GPU is collected.
https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/7b9988146881f6eee33f69c28a92ae03e2678e42/megatron/optimizer/clip_grads.py#L48-L67

The snippet below is where the gradient norm is reduced. This would refer to ![formula](https://render.githubusercontent.com/render/math?math=\mid%20G_{B_{big}}\mid^2), yes?
https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/7b9988146881f6eee33f69c28a92ae03e2678e42/megatron/optimizer/clip_grads.py#L74-L82

I suppose in order to calculate ![formula](https://render.githubusercontent.com/render/math?math=\mid%20G_{B_{small}}\mid^2)  I will have to call `total_norm = max(grad.abs().max() for grad in grads_for_norm) ` for each GPU and then calculate noise scale. Just want to verify if my hunch is correct.

@ibeltagy

**Comment by slippylolo** [comment]
Hmm, I could be wrong, but this `all_reduce` only acts across **model** parallelism. My understanding of Appendix A.1. is that this should be done across **data** parallelism instead (so actually at a higher-level than here). 

Let me dig through the code a bit to see if I can find a place where this can be accessed.

**Comment by stas00** [comment]
I think we currently don't use this, but ZeRO-DP from DeepSpeed (stage 1). I haven't verified which specific code paths it takes. It might help to step through with the debugger.

We only use TP from Megatron, but PP and DP from Deepspeed.

I probably should update the examples on how to include deepspeed zero-1. I will do that shortly.

**Comment by stas00** [comment]
Here are the instructions added: https://github.com/bigscience-workshop/Megatron-DeepSpeed#deepspeed-pp-and-zero-dp

**Comment by lintangsutawika** [comment]
So by using option `--DDP-impl torch` we can acquire the gradients from `Megatron-DeepSpeed/megatron/model/distributed.py` and can be micro batch gradient norm and the global batch gradient norm can be calculated each iteration.

I feel like even if we use Pipeline Parallelism, the difference would just be that calculating the global batch gradient norm can only be done during the backpropagation on all accumulated gradients.

What do you think @slippylolo ?

**Comment by slippylolo** [comment]
I don't think this is an acceptable solution, as ZeRO-DP comes with some pretty nice savings (plus we wouldn't want to maintain two different data parallelism scheme just for gradient noise scale measurement. 

I haven't had time to look into it yet, but there must be an easy way to acquire the local and global batch gradients in ZeRO-DP.

**Comment by lintangsutawika** [comment]
Yes, I agree. 

In that case, I think we can extract the gradients around here. ![formula](https://render.githubusercontent.com/render/math?math=%5Cmid%20G_%7BB_%7Bsmall%7D%7D%5Cmid%5E2) would be the gradients obtained from each micro-batch and ![formula](https://render.githubusercontent.com/render/math?math=%5Cmid%20G_%7BB_%7Bbig%7D%7D%5Cmid%5E2) would be the gradient after averaging all accumulations.
https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/5e3963db4a5db36e1ebdb152e2760382bdc7ef04/megatron/training.py#L441-L451

The DeepSpeed documentation explains that the gradient accumulation method is used is slightly different than what we might want. Instead of gradient reduction per micro-batch, it instead averages locally for each step and finally averages across all GPUs at the end of the accumulation sequence.
https://www.deepspeed.ai/features/#smart-gradient-accumulation
